### PR TITLE
macOS: add battery info support; fix GPU detection; various improvements

### PR DESCRIPTION
**Pull request by CarterLi**
See commit history for detail

image

### CODE REVIEW

File: src/util/FFcheckmacros.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#if defined(__GNUC__) || defined(__clang__)
+    #define FF_C_PRINTF(formatStrIndex, argsStartIndex) __attribute__((__format__ (printf, formatStrIndex, argsStartIndex)))
+#else
+    #define FF_C_PRINTF
+#endif
+
+#if defined(__GNUC__) || defined(__clang__)
+    #define FF_C_SCANF(formatStrIndex, argsStartIndex) __attribute__((__format__ (scanf, formatStrIndex, argsStartIndex)))
+#else
+    #define FF_C_SCANF
```
Comment (LinusDierheimer) [inline src/util/FFcheckmacros.h:21]:
```c
#define FF_C_SCANF(formatStrIndex, argsStartIndex)
```

File: src/util/FFcheckmacros.h [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+#if defined(__GNUC__) || defined(__clang__)
+    #define FF_C_NODISCARD __attribute__((warn_unused_result))
+#else
+    #define FF_C_NODISCARD
+#endif
+
+#if defined(__GNUC__) || defined(__clang__)
+    #define FF_C_PRINTF(formatStrIndex, argsStartIndex) __attribute__((__format__ (printf, formatStrIndex, argsStartIndex)))
+#else
+    #define FF_C_PRINTF
```
Comment (LinusDierheimer) [inline src/util/FFcheckmacros.h:15]:
```c
#define FF_C_PRINTF(formatStrIndex, argsStartIndex)
```

File: src/util/cfdict_helpers.c [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    FFstrbuf ffStr;
+    ffStrbufInitA(&ffStr, length + 1);
+    if(CFStringGetCString(cfStr, ffStr.chars, length + 1, kCFStringEncodingASCII))
+    {
+        ffStr.length = length;
+        // CFStringGetCString ensures the buffer is NUL terminated
+        // https://developer.apple.com/documentation/corefoundation/1542721-cfstringgetcstring
+    }
+    return ffStr;
+}
```
Comment (LinusDierheimer) [inline src/util/cfdict_helpers.c:22]: I'd rather have this method taking a `FFstrbuf* result` as third argument and appending to it, than returning a new one. This way the ownership of the allocated memory is clear (and the cpu has to copy less).
You can use `ffStrbufEnsureFree` to ensure enough capacity for `CFStringGetCString`

File: src/util/cfdict_helpers.c [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+int ffCfDictGetInt(CFMutableDictionaryRef dict, const char* str)
+{
+    CFNumberRef cfNum = (CFNumberRef)ffCfDictGetValue(dict, str);
+    int32_t result;
+    if(!CFNumberGetValue(cfNum, kCFNumberSInt32Type, &result))
+        return -1;
+    return result;
+}
+
+double getCfDictDouble(CFMutableDictionaryRef dict, const char* str)
```
Comment (LinusDierheimer) [inline src/util/cfdict_helpers.c:39]:
```c
double ffCfDictGetDouble(...)
```

But is this method even used anywhere?

File: src/util/cfdict_helpers.h [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+#ifndef FASTFETCH_INCLUDED_cfdict_helpers
+#define FASTFETCH_INCLUDED_cfdict_helpers
+
+#include "fastfetch.h"
+#include <CoreFoundation/CoreFoundation.h>
+
+const void* ffCfDictGetValue(CFMutableDictionaryRef dict, const char* str);
+FFstrbuf ffCfDictGetString(CFMutableDictionaryRef dict, const char* str);
+int ffCfDictGetBool(CFMutableDictionaryRef dict, const char* str);
+int ffCfDictGetInt(CFMutableDictionaryRef dict, const char* str);
```
Comment (LinusDierheimer) [inline src/util/cfdict_helpers.h:12]: Why not use `FF_C_NODISCARD` here too?

File: src/util/cfdict_helpers.c [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+int ffCfDictGetInt(CFMutableDictionaryRef dict, const char* str)
+{
+    CFNumberRef cfNum = (CFNumberRef)ffCfDictGetValue(dict, str);
+    int32_t result;
+    if(!CFNumberGetValue(cfNum, kCFNumberSInt32Type, &result))
+        return -1;
+    return result;
+}
```
Comment (LinusDierheimer) [inline src/util/cfdict_helpers.c:37]: Is it possible that we ever _want_ to read `-1`? Maybe specify an argument (default value) that is returned on error.

File: src/util/cfdict_helpers.c [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        // https://developer.apple.com/documentation/corefoundation/1542721-cfstringgetcstring
+    }
+    return ffStr;
+}
+
+int ffCfDictGetBool(CFMutableDictionaryRef dict, const char* str)
+{
+    CFBooleanRef cfNum = (CFBooleanRef)ffCfDictGetValue(dict, str);
+    return CFBooleanGetValue(cfNum);
+}
```
Comment (LinusDierheimer) [inline src/util/cfdict_helpers.c:28]: There is no error handling here. What happens if the key doesn't exist? Is false returned?

**Review by LinusDierheimer — COMMENTED** [review]
Thank you for your work on this!

File: src/util/cfdict_helpers.c (hunk 4 above)
Comment (CarterLi) [reply src/util/cfdict_helpers.c]: Removed

**Comment by LinusDierheimer** [comment]
Seems good to me, lets merge this!

**Pull request merged**